staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -692,6 +695,63 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
 	return refs, nil
 }
 
+// IsMountPoint determines if a file is a mountpoint.
+// It first detects bind & any other mountpoints using
+// MountedFast function. If the MountedFast function returns
+// sure as true and err as nil, then a mountpoint is detected
+// successfully. When an error is returned by MountedFast, the
+// following is true:
+// 1. All errors are returned with IsMountPoint as false
+// except os.IsPermission.
+// 2. When os.IsPermission is returned by MountedFast, List()
+// is called to confirm if the given file is a mountpoint are not.
+//
+// os.ErrNotExist should always be returned if a file does not exist
+// as callers have in past relied on this error and not fallback.
+//
+// When MountedFast returns sure as false and err as nil (eg: in
+// case of bindmounts on kernel version 5.10- ); mounter.List()
+// endpoint is called to enumerate all the mountpoints and check if
+// it is mountpoint match or not.
+func (mounter *Mounter) IsMountPoint(file string) (bool, error) {
+	isMnt, sure, isMntErr := mountinfo.MountedFast(file)
+	if sure && isMntErr == nil {
+		return isMnt, nil
+	}
+	// We were not allowed to do the simple stat() check, e.g. on NFS with
+	// root_squash. Fall back to /proc/mounts check below when
+	// fs.ErrPermission is returned.
+	if isMntErr != nil {
+		if errors.Is(isMntErr, fs.ErrNotExist) {
```

Comment:
also just to make sure mountinfo.MountedFast will return ErrNotExist if file/path does not exist?

looks like it returns &os.PathError{Op: "openat2", Path: dir, Err: err}, 
moby/sys@88ece65#diff-1233c1d7752b2c506417d72bf871879082263410c1ee1cf3c0021f0e47a5d4faR47

Reply:
Yep! PathError adds more context to it. That is why we are unwrapping to return bare error:

Please see this example for unwrapping cases where we will get true when errors.Is(isMntError, fs.ErrNotExist)
https://go.dev/play/p/DBNWKT8Qi6A